torch/nested/_internal/nested_tensor.py

```diff
@@ -0,0 +1,161 @@
+        self.offsets = offsets
+        self._nested_sizes = None
+        self.nb_tensors = nb_tensors
+        return
+
+    def __repr__(self):
+        # We should implement this in torch/_tensor_str.py instead
+        grad_fn_str = f", requires_grad={self.requires_grad}" if self.requires_grad else ""
+        if self.grad_fn :
+            grad_fn_str = f", grad_fn={self.grad_fn}"
+        return f"NestedTensor(size={self._size}, offsets={self.offsets}{grad_fn_str})"
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        kwargs = {} if kwargs is None else kwargs
+        any_jagged = any([is_jagged(x) for x in args])
+        any_strided = any([is_strided(x) for x in args])
+
+        if any_jagged and any_strided:
+            raise NotImplementedError
+
+        if any_jagged:
+            # Lazy import to avoid circular dependency
+            from .ops import lookup_jagged
+            fn = lookup_jagged(func, *args, **kwargs)
+            if fn is not None:
+                return fn(*args, **kwargs)
+
+        raise NotImplementedError
+
+def is_jagged(x: torch.Tensor) -> bool:
+    return isinstance(x, NestedTensor) and x.is_jagged
+
+def is_strided(x: torch.Tensor) -> bool:
+    return isinstance(x, NestedTensor) and not x.is_jagged
+
+# Not actually a view!
+class ViewBufferFromNested(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, x: NestedTensor):
+        ctx.save_for_backward(x.offsets)
+        ctx.nb_tensors = x.nb_tensors
+        return x.buffer
+
+    @staticmethod
+    def backward(ctx, gO: torch.Tensor):
+        offsets, = ctx.saved_tensors
+        nb_tensors = ctx.nb_tensors
+        return NestedTensor(gO, offsets=offsets, nb_tensors=nb_tensors)
+
+# Not actually a view!
+class ViewNestedFromBuffer(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, buffer: torch.Tensor, offsets: torch.Tensor, nb_tensors: int):
+        return NestedTensor(buffer, offsets=offsets, nb_tensors=nb_tensors)
+
+    @staticmethod
+    def backward(ctx, gO: torch.Tensor):
+        return gO.buffer, None, None
+
+# Need to make it obvious that users should be passing in offsets
+def jagged_from_list(
+        tensors: Sequence[torch.Tensor],
+        offsets: Optional[torch.Tensor]) -> Tuple[NestedTensor, torch.Tensor]:
+    """Constructs a NestedTensor backed by jagged layout from a list of tensors
+    """
+    assert len(set(t.dtype for t in tensors)) == 1
+    assert len(set(t.device for t in tensors)) == 1
+    assert all(t.ndim == 2 for t in tensors)
+    assert len(set(t.shape[1] for t in tensors)) == 1
+
+    lengths = torch.tensor([t.shape[0] for t in tensors])
+    _offsets = torch.cat([torch.tensor([0]), lengths.cumsum(0)])
+    if offsets is not None:
+        assert torch.all(offsets == _offsets).item()
+    else:
+        offsets = _offsets
+
+    return ViewNestedFromBuffer.apply(torch.cat(tensors, dim=0), offsets, len(tensors)), offsets
```

Comment:
as discussed, in-place ops on views may be broken. we talked about throwing NotImplemented for in-place ops for the initial commit